packages/core/crates/core-network/src/peers.rs

```diff
@@ -0,0 +1,360 @@
+/// Default quality for unknown or offline nodes
+const BAD_QUALITY: f64 = 0.2;
+const IGNORE_TIMEFRAME: Duration = Duration::from_secs(600); // 10 minutes
+
+// Does not work with enums
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+pub enum NetworkPeerOrigin {
+    Initialization = 0,
+    NetworkRegistry = 1,
+    IncomingConnection = 2,
+    OutgoingConnection = 3,
+    StrategyExistingChannel = 4,
+    StrategyConsideringChannel = 5,
+    StrategyNewChannel = 6,
+    ManualPing = 7,
+    Testing = 8,
+}
+
+impl std::fmt::Display for NetworkPeerOrigin {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let description = match self {
+            NetworkPeerOrigin::Initialization => "node initialization",
+            NetworkPeerOrigin::NetworkRegistry => "registered in network registry",
+            NetworkPeerOrigin::IncomingConnection => "incoming connection",
+            NetworkPeerOrigin::OutgoingConnection => "outgoing connection attempt",
+            NetworkPeerOrigin::StrategyExistingChannel => "strategy monitors existing channel",
+            NetworkPeerOrigin::StrategyConsideringChannel => "strategy considers opening a channel",
+            NetworkPeerOrigin::StrategyNewChannel => "strategy decided to open new channel",
+            NetworkPeerOrigin::ManualPing => "manual ping",
+            NetworkPeerOrigin::Testing => "testing",
+        };
+        write!(f, "{}", description)
+    }
+}
+
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+#[derive(Debug, Clone)]
+pub struct Entry {
+    id: PeerId,
+    origin: NetworkPeerOrigin,
+    last_seen: u64,     // timestamp
+    quality: f64,
+    heartbeats_sent: u64,
+    heartbeats_succeeded: u64,
+    backoff: f64,
+    ignored_at: Option<f32>,
+}
+
+impl std::fmt::Display for Entry {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Entry: [id={}, origin={}, last seen on={}, quality={}, heartbeats sent={}, heartbeats succeeded={}, backoff={}, ignored at={:#?}]",
+               self.id, self.origin, self.last_seen, self.quality, self.heartbeats_sent, self.heartbeats_succeeded, self.backoff, self.ignored_at)
+    }
+}
+
+impl Entry {
+    fn new(id: PeerId, origin: NetworkPeerOrigin) -> Entry {
+        Entry {
+            id,
+            origin,
+            heartbeats_sent: 0,
+            heartbeats_succeeded: 0,
+            last_seen: 0,
+            backoff: 2.0,
+            quality: 0.0,
+            ignored_at: None,
+        }
+    }
+
+    fn next_ping(&self) -> u64 {
+        let backoff = self.backoff.powf(BACKOFF_EXPONENT);
+        let delay = std::cmp::min(
+            MAX_DELAY,
+            Duration::from_millis((MIN_DELAY.as_millis() as f64 * backoff) as u64),
+        );
+        return self.last_seen + delay.as_millis() as u64;
+    }
+}
+
+struct NetworkPeers {
+    entries: HashMap<String, Entry>,
```

Comment:
Could we have a more declarative type for the keys of those structures?

Reply:
What do you mean? As in if PeerId could be used directly? Likely not.

Reply:
Ok. It would have been nice since that would immediately show the semantics of the key.